### ISSUE TITLE
Migration of V1 Templates to new full NP feature

### ISSUE THREAD

**Issue by jgclark** [NotePlan/plugins#232]
Templates started with work by @nmn.
@mikeerickson is about to release a major upgrade, known as Templating.
Now @EduardMe is going to make Templates a full part of the product, to reside in `@Templates` special folder.

This is the issue to discuss and track progress on migrating users' V1 templates through to the New Generation Templates (whatever that looks like).  Primary assignee is Eduard, as this appears to need the product itself to do (most of) the migration.

**Comment by jgclark** [comment]
From private Discord conversation, @EduardMe has agreed to add `NotePlan.environment.templateFolder` which will return the path to the new template folder location.  (Good.)

@mikeerickson replied:
> I will await the update whenever that may be so that "np.Templating" will be using that location to be the basePath toward template notes.

And rightly then said:
> I am envisioning a migrateTemplates type method that will move templates from the current location to the new location (wherever that may be).  I think this would be better served inside NotePlan since it has to deal with the file system, something we don't have the capability in NotePlan plugins.

**Comment by jgclark** [comment]
Hmm. Such API calls would be powerful; by accident or design they could then wipe out user data, or potentially post it out via web calls.  Though, come to think of it, we already have that ability, so scratch that thoughts.

The downside of leaving the migration to a Plugin is that it relies on users updating the Template plugin, or downloading the Templating plugin instead. And then when the migration is automatically run (by our new friend the onMigrateOrUpdate call?), somehow we disable the nmn.Template plugin. That all feels a bit risky.

**Comment by EduardMe** [comment]
@dwertheimer I agree with this. NotePlan does not know where users have their current plugins, so offering move/delete/rename is a good option.

Just to clarify, with "I don't know what I don't know" I tried to correct the expectation that if something looks simple, it doesn't mean it **is** simple. Saying something like "It would be faster to make the code change than to reply here" is not necessarily true. I ran into this mistake a few times believing naively it would be a quick change and no bugs.

Before we get into migration trouble again. The migration (if you chose to migrate) does NOT have to be ready when the next release is launched! And I think for some plugins this is now a preference, so the user would have some default already pointing to the current folder. The user could change the folder as well.

**Comment by mikeerickson** [comment]
@EduardMe So long as it works with this

```js
static async getTemplate(templateName: string = ''): Promise<string> {

    // current method, getTemplateFolder returns value from settings
    // let templateFolderName = await getTemplateFolder() // this current returns 📋 Templates

    let templateFolderName = NotePlan.environment.templateFolder
    let templateFilename = `${templateFolderName}/${templateName}.md`
    return await DataStore.projectNoteByFilename(templateFilename)
}
```

So this setting will be removed, and I will use `NotePlan.enviornment.templateFolder`
![CleanShot 2022-02-10 at 11 15 33](https://user-images.githubusercontent.com/183153/153480343-8c515227-a48c-4da5-b854-7f4baffb1576.png)

**Comment by jgclark** [comment]
I've been giving some more thought to the migration of users' templates from 📋Templates to @Templates. I'm coming round to the idea that we shouldn't offer automatic migration, and perhaps not even an on-demand migration helper. Quite apart from the complexity of providing and scheduling this, I'm hoping Templating will be released into the wild by then. It sounds like this will be different enough that it would be useful for users to think through their various templates again, not just to move from `{{...}}` syntax to the new one, but to see what new functions and parameters are available. Is this a reasonable conclusion, @mikeerickson ?

**Comment by EduardMe** [comment]
Same here @jgclark, it scares me to build in something complicated for a one-time migration. I would say, the new default value could be “@Templates” as the template folder or kept empty, in which case the `NotePlan.environment.templateFolder` value could be used as default.

There’s still the problem how existing template users would switch it over. 

We could build add a prompt that appears after updating. It could check if the user has a - now custom - template folder and suggest changing it to the new default and built-in Templates folder by removing the value in the text box and then manually moving the notes.

The plugin can check if the folder name is empty and use the `NotePlan.environment…` value instead?

**Comment by jgclark** [comment]
> Same here @jgclark, it scares me to build in something complicated for a one-time migration. I would say, the new default value could be “@templates” as the template folder or kept empty, in which case the `NotePlan.environment.templateFolder` value could be used as default.
> 
> There’s still the problem how existing template users would switch it over.

This is more difficult than it first seems, isn't it?

We have 4 things that can help us:
1. All templates are currently held in a single folder (unless someone is seriously rolling their own).
2. The new @Templates is going to be visible in the UI (yes?) so it will be easier to ask people to shift things to it.
3. We could update the current /...Template commands to put up a warning each time they're used, to encourage people to move them.

4. What you say next:
> We could build add a prompt that appears after updating. It could check if the user has a - now custom - template folder and suggest changing it to the new default and built-in Templates folder by removing the value in the text box and then manually moving the notes.

How many ways will we have of using templates soon? Is it just the existing plugin commands, and @mikeerickson's new Templating? @EduardMe are you going to be adding a more general-purpose way of invoking templates, or just meeting notes?

@mikeerickson: does Templating have its own equivalent of /insertTemplate and /applyTemplate? If so is there a way we could get them to be used in preference to the older ones if both plugins are installed?

> The plugin can check if the folder name is empty and use the `NotePlan.environment…` value instead?

I'm still not sure in all this when/how is the best way to switch to setting `NotePlan.environment.templateFolder` to `@Templates`.

**Comment by EduardMe** [comment]
> This is more difficult than it first seems, isn't it?
I'm sure it is! But I'm trying to simplify as much as we can.

> The new @templates is going to be visible in the UI (yes?) so it will be easier to ask people to shift things to it.
Excatly, the use can move them over by drag & drop:

<img width="224" alt="image" src="https://user-images.githubusercontent.com/3285090/154067185-c1793c3b-95b8-4c50-b4a0-d15b76d179be.png">

> We could update the current /...Template commands to put up a warning each time they're used, to encourage people to move them.
I can imagine this can get very annoying for the user. That's why I proposed to show it only after updating. The user would also eventually see the template folder and be encouraged to move the notes. I think this one is not as critical as the V1 config. A custom templates folder is not fragile, just not very elegant (a necessary workaround).

> How many ways will we have of using templates soon? Is it just the existing plugin commands, and @mikeerickson's new Templating? @EduardMe are you going to be adding a more general-purpose way of invoking templates, or just meeting notes?

At the moment this is very specifically for Meeting Notes due to the frontmatter behavior (like where the meeting note should be created) and I want to embed the event into the note somehow. 

For the normal note templates, I can imagine we can later ship the template plugin inside NotePlan (like the plugin update) and add hooks or buttons, so it's a little easier to apply templates. For example if a new daily note is empty, it could contain a "apply daily note template..." button. 

> I'm still not sure in all this when/how is the best way to switch to setting NotePlan.environment.templateFolder to @Templates.

We can use 3.4.1 beta for testing and 3.5 to pull the trigger. I have built-in only the folder and detecting front matter. But it needs some bug fixing. For the 3.4.1 release, I'll hide the folder if it's empty for now and make it visible with 3.5.

**Comment by jgclark** [comment]
I'm happy to declare this feature request well and truly met, thanks to sterling work by @mikeerickson and @EduardMe.
